hazelcast/src/main/java/com/hazelcast/internal/nio/Packet.java

```diff
@@ -330,27 +335,53 @@ public String describeFlags(char flags) {
          * <p>
          * {@code ordinal = 7}
          */
-        UNDEFINED7;
+        UNDEFINED7,
+        /**
+         * Type reserved for compatibility (3.x) bind messages. The appropriate
+         * type conversion will happen in {@link #fromFlags(int, boolean)}.
+         * <p>
+         * {@code ordinal = 4}
+         */
+        COMPATIBILITY_BIND_MESSAGE(4),
+        /**
+         * Type reserved for compatibility (3.x) extended bind messages. The appropriate
+         * type conversion will happen in {@link #fromFlags(int, boolean)}.
+         * <p>
+         * {@code ordinal = 5}
+         */
+        COMPATIBILITY_EXTENDED_BIND(5);
 
         final char headerEncoding;
 
         private static final Type[] VALUES = values();
 
         Type() {
-            headerEncoding = (char) encodeOrdinal();
+            headerEncoding = (char) encodeHeader(ordinal());
         }
 
-        public static Type fromFlags(int flags) {
-            return VALUES[headerDecode(flags)];
+        Type(int ordinal) {
+            headerEncoding = (char) encodeHeader(ordinal);
+        }
+
+        public static Type fromFlags(int flags, boolean isCompatibility) {
+            int ordinal = headerDecode(flags);
+            if (isCompatibility) {
+                if (ordinal == SQL.ordinal()) {
```

Comment:
Looks good, maybe we can add a constant instead of using `SQL`'s ordinal to avoid confusion.

Reply:
Added, with a comment :)